### PR TITLE
Integrate Peach Mutation strategy

### PR DESCRIPTION
**Pull request by mpherman2**

### CODE REVIEW

File: src/python/bot/fuzzers/libFuzzer/peach/pits.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+from system import environment
+import os
+
+PIT_DIR = os.path.join(environment.get_platform_resources_directory(), 'peach',
+                       'pits')
+PATH = 0
+TITLE = 1
+# The key is the name of the grammar as defined in the libfuzzer options file.
+# The value is the path to the pit, and the title of the pit.
+PIT_INFORMATION = {'PDF': (os.path.join(PIT_DIR, 'pdf.xml'), 'PDF')}
```
Comment (mbarbella-chromium) [inline src/python/bot/fuzzers/libFuzzer/peach/pits.py]: Having this mapping seems like overkill to me. I'd prefer that get path function just adds ".xml" to its argument and returns None if the file doesn't exist. Callers could check this and log error so that we know if things are improperly configured.

File: src/python/bot/fuzzers/libfuzzer.py [hunk 2]
```diff
@@ -49,6 +51,7 @@
 from system import minijail
 from system import new_process
 from system import shell
+from zipfile import ZipFile
```
Comment (mbarbella-chromium) [inline src/python/bot/fuzzers/libfuzzer.py]: Nit: instead of this, "from system import archive" and use archive.unpack instead. It's fine to pass trusted=True given that the file is checked in.

File: src/python/bot/fuzzers/libfuzzer.py [hunk 3]
```diff
@@ -1589,6 +1592,55 @@ def use_radamsa_mutator_plugin(extra_env):
...
+  new_path = new_path + sys.path
+  extra_env['PYTHONPATH'] = ':'.join(new_path)
+
+  # Select grammar.
+  grammar = fuzzer_utils.extract_argument(arguments, 'grammar')
+  # If the target has no grammar, do not use this strategy.
+  if not grammar:
+    return False
+
+  if not pits.validate(grammar):
```
Comment (mbarbella-chromium) [inline src/python/bot/fuzzers/libfuzzer.py]: Can we do this check earlier, ideally before extracting?

File: src/python/bot/fuzzers/libFuzzer/peach/pits.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def validate(grammar):
```
Comment (mbarbella-chromium) [inline src/python/bot/fuzzers/libFuzzer/peach/pits.py]: This and get_title could be removed if the comment above is addressed. The return value of get_path would be used for validation.

File: src/python/fuzzing/strategy.py [hunk 5]
```diff
@@ -46,6 +46,8 @@
     name='recommended_dict', probability=0.10, manually_enable=False)
 VALUE_PROFILE_STRATEGY = Strategy(
     name='value_profile', probability=0.33, manually_enable=False)
+PEACH_GRAMMAR_MUTATION_STRATEGY = Strategy(
+    name='peach_grammar_mutation', probability=0.10, manually_enable=True)
```
Comment (mbarbella-chromium) [inline src/python/fuzzing/strategy.py]: This is fine to start, but we should likely increase it later. Since relatively few jobs will support this, we'll want to run it more often so that it's actually used when it can be used.

File: src/python/bot/fuzzers/libFuzzer/peach/pits.py (hunk 1 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/libFuzzer/peach/pits.py]: Done

File: src/python/bot/fuzzers/libFuzzer/peach/pits.py (hunk 4 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/libFuzzer/peach/pits.py]: Done

File: src/python/bot/fuzzers/libfuzzer.py (hunk 2 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/libfuzzer.py]: Done

File: src/python/bot/fuzzers/libfuzzer.py (hunk 3 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/libfuzzer.py]: Done

File: src/python/fuzzing/strategy.py (hunk 5 above)
Comment (mpherman2) [reply src/python/fuzzing/strategy.py]: Ack

File: src/python/bot/fuzzers/libfuzzer.py [hunk 6]
```diff
@@ -1589,6 +1592,58 @@ def use_radamsa_mutator_plugin(extra_env):
...
+  """Decide whether or not to use peach mutator, and set up all of the
+  environment variables necessary to do so."""
+  # TODO(mpherman): Include architecture info in job definition and exclude
+  #  i386.
+  if environment.platform() != 'LINUX' or environment.get_value(
+      'MEMORY_TOOL') != 'ASAN':
+    return False
+
+  # Select grammar.
+  grammar = fuzzer_utils.extract_argument(arguments, 'grammar')
```
Comment (mbarbella-chromium) [inline src/python/bot/fuzzers/libfuzzer.py]: Could we move more of this logic into get_path? Essentially everything from here to through the log_error, returning None instead of false, then returning False from this function if not pit_path.

File: src/python/bot/fuzzers/libfuzzer.py [hunk 7]
```diff
@@ -1589,6 +1592,58 @@ def use_radamsa_mutator_plugin(extra_env):
...
+  source = os.path.join(peach_dir, 'peach_mutator.zip')
+
+  archive.unpack(source, unzipped, trusted=True)
+
+  # Set LD_PRELOAD.
+  peach_path = os.path.join(unzipped, 'peach_mutator', 'src', 'peach.so')
+  extra_env['LD_PRELOAD'] = peach_path
+
+  # Set Python path.
+  # TODO(mpherman): Change this to explicitly point to Python 2
```
Comment (mbarbella-chromium) [inline src/python/bot/fuzzers/libfuzzer.py]: Nit: end with period.

File: src/python/fuzzing/strategy.py (hunk 5 above)
Comment (mbarbella-chromium) [reply src/python/fuzzing/strategy.py]: Thinking more, could you add a TODO here so we don't forget about that? Sorry for not mentioning it last time.

File: src/python/bot/fuzzers/libfuzzer.py (hunk 6 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/libfuzzer.py]: Done

File: src/python/bot/fuzzers/libfuzzer.py (hunk 7 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/libfuzzer.py]: Done

File: src/python/fuzzing/strategy.py (hunk 5 above)
Comment (mpherman2) [reply src/python/fuzzing/strategy.py]: Done

File: src/python/bot/fuzzers/libfuzzer.py [hunk 8]
```diff
@@ -1589,6 +1592,52 @@ def use_radamsa_mutator_plugin(extra_env):
   return True
 
 
+def use_peach_mutator(extra_env, arguments):
+  """Decide whether or not to use peach mutator, and set up all of the
+  environment variables necessary to do so."""
+  # TODO(mpherman): Include architecture info in job definition and exclude
+  #  i386.
+  if environment.platform() != 'LINUX' or environment.get_value(
```
Comment (oliverchang) [inline src/python/bot/fuzzers/libfuzzer.py]: nit: This line continuation looks a bit strange.

I'd do:

```python
if (environment.platform() != 'LINUX' or
    environment.get_value('MEMORY_TOOL') != 'ASAN'):
  ...

File: src/python/bot/fuzzers/libfuzzer.py [hunk 9]
```diff
@@ -1589,6 +1592,52 @@ def use_radamsa_mutator_plugin(extra_env):
...
+
+  archive.unpack(source, unzipped, trusted=True)
+
+  # Set LD_PRELOAD.
+  peach_path = os.path.join(unzipped, 'peach_mutator', 'src', 'peach.so')
+  extra_env['LD_PRELOAD'] = peach_path
+
+  # Set Python path.
+  # TODO(mpherman): Change this to explicitly point to Python 2.
+  new_path = []
```
Comment (oliverchang) [inline src/python/bot/fuzzers/libfuzzer.py]: nit: just do

```python
new_path = [
  os.path.join(unzipped, 'peach_mutator', 'src'),
  os.path.join(unzipped, 'peach_mutator', 'third_party', 'peach')
] + sys.path
```

File: src/python/bot/fuzzers/libfuzzer.py [hunk 10]
```diff
@@ -1589,6 +1592,52 @@ def use_radamsa_mutator_plugin(extra_env):
...
+  new_path.append(os.path.join(unzipped, 'peach_mutator', 'src'))
+  new_path.append(
+      os.path.join(unzipped, 'peach_mutator', 'third_party', 'peach'))
+
+  # Append the mutator to the begining of the Python path.
+  new_path = new_path + sys.path
+  extra_env['PYTHONPATH'] = ':'.join(new_path)
```
Comment (oliverchang) [inline src/python/bot/fuzzers/libfuzzer.py]: us os.pathsep here.

File: src/python/bot/fuzzers/libfuzzer.py [hunk 11]
```diff
@@ -1589,6 +1592,52 @@ def use_radamsa_mutator_plugin(extra_env):
...
+  """Decide whether or not to use peach mutator, and set up all of the
+  environment variables necessary to do so."""
+  # TODO(mpherman): Include architecture info in job definition and exclude
+  #  i386.
+  if environment.platform() != 'LINUX' or environment.get_value(
+      'MEMORY_TOOL') != 'ASAN':
+    return False
+
+  # Select grammar.
+  grammar = fuzzer_utils.extract_argument(arguments, 'grammar')
```
Comment (oliverchang) [inline src/python/bot/fuzzers/libfuzzer.py]: How is the grammar piped to here? I.e. how is the grammar specified for a target such that at this point there is an argument here?

Also, we shouldn't be using `arguments` for these (the usage seems wrong anyway, as we need to specify the argument of the format `-flag=`). That should be libFuzzer-specific arguments. If we want to add additional metadata, we can add a field to StrategyInfo.

File: src/python/bot/fuzzers/libfuzzer.py (hunk 8 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/libfuzzer.py]: Done

File: src/python/bot/fuzzers/libfuzzer.py (hunk 9 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/libfuzzer.py]: Done

File: src/python/bot/fuzzers/libfuzzer.py (hunk 10 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/libfuzzer.py]: Done

File: src/python/bot/fuzzers/libfuzzer.py (hunk 11 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/libfuzzer.py]: We add this "grammar=..." in the libfuzzer options. This it how it ends up in the arguments. Is this not how we should do it?

Comment (oliverchang) [reply src/python/bot/fuzzers/libfuzzer.py]: IMO we should create a new section in .options for grammar specific stuff, or even an entirely new file like TARGET.grammar or something similar.

We shouldn't combine libFuzzer specific and non-libFuzzer specific options under the one `arguments`, as that creates confusion.

@mbarbella-chromium WDYT?

Comment (oliverchang) [reply src/python/bot/fuzzers/libfuzzer.py]: Another reason not to do this is if we don't choose peach strategy, this argument will be passed to the libFuzzer target, which is wrong and can potentially break things.

Again, we can avoid all of this confusion if we pipe it through via a different mechanism.

Comment (mbarbella-chromium) [reply src/python/bot/fuzzers/libfuzzer.py]: Good point. We definitely don't want to be passing this along if we aren't picking this strategy. Having the separate .grammar file sounds like the cleanest solution to me.

File: src/python/bot/fuzzers/libFuzzer/peach/pits.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Dictionary to keep track of pit information."""
+
+from metrics import logs
+from system import environment
+import os
```
Comment (Dor1s) [inline src/python/bot/fuzzers/libFuzzer/peach/pits.py]: nit: move this system import above the local ones

File: src/python/bot/fuzzers/libFuzzer/peach/pits.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+
+def get_path(grammar):
+  """Return the path of the peach pit for the given grammar. Return None if the
+  Pit does not exist or the grammar is None."""
```
Comment (Dor1s) [inline src/python/bot/fuzzers/libFuzzer/peach/pits.py]: do you anticipate it ever being called with an empty `grammar`? I'd rather make sure on the caller side that the grammar is always non-empty and then log an error here if it ever gets call with an empty argument. Otherwise, all these cases will go silent and may hide some other issue from us

File: src/python/bot/fuzzers/libFuzzer/peach/pits.py [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+  # If the target has no grammar, do not use this strategy.
+  if not grammar:
+    return None
+
+  pit_dir = os.path.join(environment.get_platform_resources_directory(),
+                         'peach', 'pits')
+  pit_path = os.path.join(pit_dir, grammar + '.xml')
+
+  if not os.path.exists(pit_path):
+    logs.log_error('Peach Mutator is being used with the grammar named ' +
```
Comment (Dor1s) [inline src/python/bot/fuzzers/libFuzzer/peach/pits.py]: nit: I'd write `'Pit file for "%s" grammar is not found.' % grammar`

File: src/python/bot/fuzzers/libfuzzer.py [hunk 15]
```diff
@@ -1589,6 +1592,50 @@ def use_radamsa_mutator_plugin(extra_env):
   return True
 
 
+def use_peach_mutator(extra_env, arguments):
+  """Decide whether or not to use peach mutator, and set up all of the
+  environment variables necessary to do so."""
+  # TODO(mpherman): Include architecture info in job definition and exclude
+  #  i386.
```
Comment (Dor1s) [inline src/python/bot/fuzzers/libfuzzer.py]: super nit: extra space after `#`

Comment (Dor1s) [reply src/python/bot/fuzzers/libfuzzer.py]: seems to be copied from 1582, please fix there too :)

File: src/python/bot/fuzzers/libfuzzer.py (hunk 8 above)
Comment (Dor1s) [reply src/python/bot/fuzzers/libfuzzer.py]: why do we restrict peach mutator to this configuration only?

File: src/python/metrics/fuzzer_stats_schema.py [hunk 16]
```diff
@@ -363,6 +363,10 @@
     'mode': 'NULLABLE',
     'name': 'strategy_mutator_plugin_radamsa',
     'type': 'INTEGER'
+}, {
+    'mode': 'NULLABLE',
+    'name': 'strategy_peach_grammar_mutation',
+    'type': 'INTEGER'
```
Comment (Dor1s) [inline src/python/metrics/fuzzer_stats_schema.py]: would it make sense to record name of the grammar used?

Comment (mbarbella-chromium) [reply src/python/metrics/fuzzer_stats_schema.py]: +1, that makes more sense as the value here.

**Comment by mpherman2** [comment]
PTAL. I pushed a first draft of clusterfuzz side changes to transition to a grammar file instead of using the options file. The code still needs some unit tests, and I would like to run it locally as well, but I wanted to see if anyone had any high level comments.

File: src/python/bot/fuzzers/options.py [hunk 17]
```diff
@@ -157,22 +158,59 @@ def get_ubsan_options(self):
...
+    return None
+
+  config = configparser.ConfigParser()
+  with open(grammar_file_path, 'r') as f:
+    try:
+      config.read_file(f)
+    except configparser.Error:
+      raise FuzzerOptionsException('Failed to parse fuzzer grammar file.')
+
+  if not config.has_section('grammar'):
```
Comment (oliverchang) [inline src/python/bot/fuzzers/options.py]: this seems fine, but any reason why we want to complicate this with a config format, rather than say a plain string? Do we anticipate that we need to attach more metadata here?

File: src/python/bot/fuzzers/libfuzzer.py (hunk 8 above)
Comment (oliverchang) [reply src/python/bot/fuzzers/libfuzzer.py]: Looks like you updated lines 1582-1584, but not here. We should move this into a shared function and call it from both places.

Comment (mpherman2) [reply src/python/bot/fuzzers/libfuzzer.py]: Done

File: src/python/bot/fuzzers/options.py (hunk 17 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/options.py]: Not presently, but it is likely in the future it will be helpful to have a place to put more grammar metadata. All of the infrastructure was there already anyway, so might as well make our lives easier in the future.

File: src/python/tests/core/bot/fuzzers/libFuzzer/engine_test.py [hunk 18]
```diff
@@ -344,6 +344,7 @@ def mock_merge(*args, **kwargs):  # pylint: disable=unused-argument
         'strategy_fork': 0,
         'strategy_mutator_plugin': 0,
         'strategy_mutator_plugin_radamsa': 0,
+        'strategy_peach_grammar_mutation': 0,
```
Comment (oliverchang) [inline src/python/tests/core/bot/fuzzers/libFuzzer/engine_test.py]: The type defined in fuzzer_stats_schema is a STRING, while this is an int?

File: src/python/fuzzing/strategy.py [hunk 19]
```diff
@@ -72,6 +76,7 @@
 LIBFUZZER_STRATEGIES_WITH_PREFIX_VALUE = [
     CORPUS_SUBSET_STRATEGY,
     FORK_STRATEGY,
+    PEACH_GRAMMAR_MUTATION_STRATEGY,
```
Comment (oliverchang) [inline src/python/fuzzing/strategy.py]: I don't think this will work as expected unfortunately. 

This (https://github.com/google/clusterfuzz/blob/5e9b02204428cb1a81993d512844966d2bee0a80/src/python/bot/fuzzers/libFuzzer/stats.py#L148) assumes that the part after the prefix is an int, which is not the case here.

We need to do some refactoring to support this. While the ideal refactor is a bit complicated (I can do this later), to unblock you you can probably just refactor LIBFUZZER_STRATEGIES_WITH_PREFIX_VALUE such that it's a dict where the values indicate the type.

File: src/python/tests/core/bot/fuzzers/libFuzzer/engine_test.py (hunk 18 above)
Comment (mpherman2) [reply src/python/tests/core/bot/fuzzers/libFuzzer/engine_test.py]: I am not sure where this value is being set. It seems to be 0 by default. Do you know what I need to change to make this the tag?

File: src/python/fuzzing/strategy.py (hunk 19 above)
Comment (mpherman2) [reply src/python/fuzzing/strategy.py]: Changing LIBFUZZER_STRATEGIES_WITH_PREFIX_VALUE to a dict is not going to work because all that is available is the strategy.name, not the strategy itself. Instead I can make another dict that maps strategy.name -> type. Is that sufficient?

Comment (oliverchang) [reply src/python/fuzzing/strategy.py]: That also works.

File: src/python/tests/core/bot/fuzzers/libFuzzer/engine_test.py (hunk 18 above)
Comment (oliverchang) [reply src/python/tests/core/bot/fuzzers/libFuzzer/engine_test.py]: Most likely

https://github.com/google/clusterfuzz/blob/5e9b02204428cb1a81993d512844966d2bee0a80/src/python/bot/fuzzers/libFuzzer/stats.py#L205

File: src/python/fuzzing/strategy.py (hunk 19 above)
Comment (mpherman2) [reply src/python/fuzzing/strategy.py]: Done

File: src/python/bot/fuzzers/libFuzzer/stats.py [hunk 20]
```diff
@@ -202,7 +203,10 @@ def parse_performance_features(log_lines, strategies, arguments):
 
   # Initialize all strategy stats as disabled by default.
   for strategy_type in strategy.LIBFUZZER_STRATEGY_LIST:
-    stats[strategy_column_name(strategy_type.name)] = 0
+    if strategy_type.name == 'peach_grammar_mutation':
```
Comment (oliverchang) [inline src/python/bot/fuzzers/libFuzzer/stats.py]: This is a bit too hacky.  Let's use STRATEGIES_WITH_PREFIX_VALUE_TYPE here to decide the default value.

Comment (mpherman2) [reply src/python/bot/fuzzers/libFuzzer/stats.py]: Let me know if this is what you had in mind. The other way I considered doing it was to store the default value of the prefix in the STRATEGIES_WITH_PREFIX_VALUE_TYPE dict as well. Which would you prefer?

File: src/python/bot/fuzzers/libfuzzer.py [hunk 21]
```diff
@@ -1589,6 +1597,50 @@ def use_radamsa_mutator_plugin(extra_env):
...
+
+  # Set title and pit environment variables
+  extra_env['PIT_PATH'] = pit_path
+  extra_env['PIT_TITLE'] = grammar
+
+  # Extract zip of peach mutator code.
+  peach_dir = os.path.join(environment.get_platform_resources_directory(),
+                           'peach')
+  unzipped = os.path.join(peach_dir, 'mutator')
+  source = os.path.join(peach_dir, 'peach_mutator.zip')
```
Comment (Dor1s) [inline src/python/bot/fuzzers/libfuzzer.py]: why do we keep it in zip and extract every time?

File: src/python/bot/fuzzers/options.py [hunk 22]
```diff
@@ -157,22 +158,59 @@ def get_ubsan_options(self):
     return self._get_option_section('ubsan')
 
 
+def get_fuzz_target_support_file(fuzz_target_path, file_extension):
+  """Helper function. returns a file path for a fuzz target support file
```
Comment (Dor1s) [inline src/python/bot/fuzzers/options.py]: nit: capitalize "r" in "returns"

Comment (Dor1s) [reply src/python/bot/fuzzers/options.py]: also can skip "Helper function. "

File: src/python/bot/fuzzers/options.py [hunk 23]
```diff
@@ -157,22 +158,59 @@ def get_ubsan_options(self):
...
+  that has the given file extension"""
```
Comment (Dor1s) [inline src/python/bot/fuzzers/options.py]: and add trailing period

File: src/python/fuzzing/strategy.py [hunk 24]
```diff
@@ -72,8 +77,15 @@
...
 ]
 
+STRATEGIES_WITH_PREFIX_VALUE_TYPE = {
```
Comment (Dor1s) [inline src/python/fuzzing/strategy.py]: nit: append `LIBFUZZER_` to the variable name for consistency

File: src/python/bot/fuzzers/options.py [hunk 25]
```diff
@@ -157,22 +158,59 @@ def get_ubsan_options(self):
...
+
   try:
     return FuzzerOptions(options_file_path, cwd=options_cwd)
   except FuzzerOptionsException:
     logs.log_error('Invalid options file: %s.' % options_file_path)
     return None
+
+
+def get_fuzz_target_grammar(fuzz_target_path):
+  """Return the grammra for a given target or None if it does not exist"""
```
Comment (Dor1s) [inline src/python/bot/fuzzers/options.py]: typo in `grammra` and missing trailing period

File: src/python/bot/fuzzers/options.py (hunk 17 above)
Comment (Dor1s) [reply src/python/bot/fuzzers/options.py]: yeah, this confused me in Chromium side CL as well, I don't see why we need a `.grammar` file at all if we just need to pass an option that goes into a separate section of the `.option` file

File: src/python/bot/fuzzers/libfuzzer.py (hunk 11 above)
Comment (Dor1s) [reply src/python/bot/fuzzers/libfuzzer.py]: I frankly prefer having a new section in the `.options` file instead.
- We already have there support for libFuzzer, AFL, env variables and various sanitizer options.
- That would also keeps things a bit more simple for OSS-Fuzz, as some projects use `.options` files.
- On Chromium side we will re-use the same script for generating `.options` file instead of duplicating it.

Comment (mpherman2) [reply src/python/bot/fuzzers/libfuzzer.py]: @mbarbella-chromium @oliverchang Thoughts?

Comment (mbarbella-chromium) [reply src/python/bot/fuzzers/libfuzzer.py]: A new section seems fine to me as well.

Would this be the first time we're using a new section for something other than arguments? If so, what happens if a header is omitted? I just want to confirm that we won't suddenly have issues if any OSS-Fuzz targets had improperly formed options files, but I suspect the risk would be very low.

File: src/python/bot/fuzzers/libfuzzer.py (hunk 21 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/libfuzzer.py]: Because the peach code has to be python 2, and we don't want python 2 code in the repo. @mbarbella-chromium and I figure this would be the best way around that.

File: src/python/bot/fuzzers/options.py (hunk 22 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/options.py]: Done

File: src/python/bot/fuzzers/options.py (hunk 23 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/options.py]: Done

File: src/python/bot/fuzzers/options.py (hunk 25 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/options.py]: Done

File: src/python/fuzzing/strategy.py (hunk 24 above)
Comment (mpherman2) [reply src/python/fuzzing/strategy.py]: Done

File: src/python/bot/fuzzers/libfuzzer.py (hunk 11 above)
Comment (Dor1s) [reply src/python/bot/fuzzers/libfuzzer.py]: > Would this be the first time we're using a new section for something other than arguments?

No, only `[libfuzzer]` and `[AFL]` sections' values go into arguments.

`[env]` elements go into `VARIABLE=VALUE` env vars

`[asan]` and other sanitizers sections go into `ASAN_OPTIONS=VARIABLE=VALUE:...` and such

I think these are all different use cases, and having something new isn't harmful.

> If so, what happens if a header is omitted? 

what do you mean by "header"?

**Comment by mpherman2** [comment]
I changed getting the grammar to use an environment variable. We should just be able to add LIBFUZZER_GRAMMAR to the environment variables section of the options file without any extra changes needed to chromium. PTAL.

File: src/python/bot/fuzzers/options.py [hunk 26]
```diff
@@ -31,6 +31,7 @@
 from system import environment
 
 OPTIONS_FILE_EXTENSION = '.options'
+GRAMMAR_FILE_EXTENSION = '.grammar'
```
Comment (mbarbella-chromium) [inline src/python/bot/fuzzers/options.py]: This should be unused now, right?

File: src/python/bot/fuzzers/options.py [hunk 27]
```diff
@@ -157,20 +158,32 @@ def get_ubsan_options(self):
     return self._get_option_section('ubsan')
 
 
+def get_fuzz_target_support_file(fuzz_target_path, file_extension):
```
Comment (mbarbella-chromium) [inline src/python/bot/fuzzers/options.py]: It's probably best to inline this back into the other function. I suspect we won't be adding new files later based on the discussions in this PR.

File: src/python/bot/fuzzers/options.py (hunk 26 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/options.py]: Yes. Removed

File: src/python/bot/fuzzers/options.py (hunk 27 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/options.py]: Done

File: src/python/bot/fuzzers/libFuzzer/peach/pits.py [hunk 28]
```diff
@@ -0,0 +1,34 @@
...
+def get_path(grammar):
+  """Return the path of the peach pit for the given grammar. Return None if the
+  Pit does not exist or the grammar is None."""
+
+  pit_dir = os.path.join(environment.get_platform_resources_directory(),
+                         'peach', 'pits')
+  pit_path = os.path.join(pit_dir, grammar + '.xml')
+
+  if not os.path.exists(pit_path):
+    logs.log_error('Pit file for "%s" grammar is not found.' % grammar)
```
Comment (Dor1s) [inline src/python/bot/fuzzers/libFuzzer/peach/pits.py]: can you also log `pit_path`? I think the following should work:

```python
  if not os.path.exists(pit_path):
    logs.log_error('Pit file for "%s" grammar is not found.' % grammar,
                   pit_path=pit_path)
```

and the value will be included in the log metadata

File: src/python/bot/fuzzers/libFuzzer/stats.py [hunk 29]
```diff
@@ -144,8 +144,10 @@ def parse_line_for_strategy_prefix(line, strategy_name):
     if not line.startswith(strategy_prefix):
       return
 
+    prefix_type = strategy.LIBFUZZER_STRATEGIES_WITH_PREFIX_VALUE_TYPE[
```
Comment (Dor1s) [inline src/python/bot/fuzzers/libFuzzer/stats.py]: it's `suffix_type` I believe, that prefix is always a string like `corpus_subset_` followed by a value we parse here

File: src/python/bot/fuzzers/libFuzzer/stats.py [hunk 30]
```diff
@@ -202,7 +204,13 @@ def parse_performance_features(log_lines, strategies, arguments):
 
   # Initialize all strategy stats as disabled by default.
   for strategy_type in strategy.LIBFUZZER_STRATEGY_LIST:
-    stats[strategy_column_name(strategy_type.name)] = 0
+    if (strategy_type.name in
+        strategy.LIBFUZZER_STRATEGIES_WITH_PREFIX_VALUE_TYPE and
+        strategy.LIBFUZZER_STRATEGIES_WITH_PREFIX_VALUE_TYPE[strategy_type.name]
+        == str):
+      stats[strategy_column_name(strategy_type.name)] = 'None'
```
Comment (Dor1s) [inline src/python/bot/fuzzers/libFuzzer/stats.py]: @oliverchang is there any different whether we use `'None'` or an empty string, WDYT? I definitely don't want `None` (aka `NULL`) as that should indicate that strategy didn't existed at the time of a particular run

File: src/python/bot/fuzzers/libfuzzer.py (hunk 21 above)
Comment (Dor1s) [reply src/python/bot/fuzzers/libfuzzer.py]: ack

File: src/python/bot/fuzzers/libFuzzer/peach/pits.py (hunk 28 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/libFuzzer/peach/pits.py]: Done

File: src/python/bot/fuzzers/libFuzzer/stats.py (hunk 29 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/libFuzzer/stats.py]: Done

File: src/python/bot/fuzzers/libFuzzer/stats.py (hunk 30 above)
Comment (oliverchang) [reply src/python/bot/fuzzers/libFuzzer/stats.py]: Good point, empty string may be better here. 'None' shouldn't result in a NULL value, but empty ('') is still cleaner as the "default" value for a string.

File: src/python/bot/fuzzers/options.py [hunk 31]
```diff
@@ -160,19 +160,22 @@ def get_ubsan_options(self):
 def get_fuzz_target_options(fuzz_target_path):
   """Return a FuzzerOptions for the given target, or None if it does not
   exist."""
-  options_file_path = fuzzer_utils.get_supporting_file(fuzz_target_path,
-                                                       OPTIONS_FILE_EXTENSION)
-  options_cwd = os.path.dirname(options_file_path)
+  file_path = fuzzer_utils.get_supporting_file(fuzz_target_path,
```
Comment (oliverchang) [inline src/python/bot/fuzzers/options.py]: What's the reason for the changes here? It seems like we're just effectively renaming a variable? I prefer the old one tbh.

File: src/python/bot/fuzzers/libFuzzer/engine.py [hunk 32]
```diff
@@ -101,12 +101,13 @@ def prepare(self, corpus_dir, target_path, _):
       A FuzzOptions object.
     """
     arguments = fuzzer.get_arguments(target_path)
+    grammar = environment.get_value('LIBFUZZER_GRAMMAR', None)
```
Comment (oliverchang) [inline src/python/bot/fuzzers/libFuzzer/engine.py]: Why was this changed to an env var? 

There are a number of issues with this:
1. This doesn't actually propagated anywhere for libFuzzer. There is a [whitelist](https://github.com/google/clusterfuzz/blob/5e9b02204428cb1a81993d512844966d2bee0a80/src/python/bot/fuzzers/options.py#L36), and the propagation currently only happens for [AFL](https://github.com/google/clusterfuzz/search?q=get_env&unscoped_q=get_env).

2. Using env vars for this seems a bit hacky -- env vars are very general, pollute the namespace of multiple processes, and we should avoid adding any more to our (already very large list) if we can.. An explicit section in the options file (or the .grammar) is better.

File: src/python/bot/fuzzers/libFuzzer/stats.py [hunk 33]
```diff
@@ -202,7 +204,13 @@ def parse_performance_features(log_lines, strategies, arguments):
 
   # Initialize all strategy stats as disabled by default.
   for strategy_type in strategy.LIBFUZZER_STRATEGY_LIST:
-    stats[strategy_column_name(strategy_type.name)] = 0
+    if (strategy_type.name in
+        strategy.LIBFUZZER_STRATEGIES_WITH_PREFIX_VALUE_TYPE and
+        strategy.LIBFUZZER_STRATEGIES_WITH_PREFIX_VALUE_TYPE[strategy_type.name]
```
Comment (oliverchang) [inline src/python/bot/fuzzers/libFuzzer/stats.py]: nit: just do

```python
strategy.LIBFUZZER_STRATEGIES_WITH_PREFIX_VALUE_TYPE.get(strategy_type.name) == str
```

To avoid the "in" check.

File: src/python/bot/fuzzers/libFuzzer/engine.py (hunk 32 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/libFuzzer/engine.py]: We thought that it would get piped in from the environment variables section of the libfuzzer options file without any need for a change on the chromium side. These are good points though, so now it works with a separate grammar section in the libfuzzer options file.

File: src/python/bot/fuzzers/options.py [hunk 34]
```diff
@@ -171,6 +174,11 @@ def get_fuzz_target_options(fuzz_target_path):
   if not os.path.exists(options_file_path):
     return None
 
+  if not options_file_path:
```
Comment (oliverchang) [inline src/python/bot/fuzzers/options.py]: nit: move this to immediately after line 167-168.

File: src/python/bot/fuzzers/libfuzzer.py [hunk 35]
```diff
@@ -1708,6 +1763,14 @@ def pick_strategies(strategy_pool, fuzzer_path, corpus_directory,
     fuzzing_strategies.append(strategy.MUTATOR_PLUGIN_STRATEGY.name)
 
   if (strategy.MUTATOR_PLUGIN_STRATEGY.name not in fuzzing_strategies and
+      strategy_pool.do_strategy(strategy.PEACH_GRAMMAR_MUTATION_STRATEGY) and
```
Comment (oliverchang) [inline src/python/bot/fuzzers/libfuzzer.py]: nit: this is getting a bit difficult to read and will get even more so if we add more.

maybe do something like

```python
MUTATOR_STRATAGIES = [
  strategy.PEACH_GRAMMAR_MUTATION_STRATEGY.name,
  strategy.MUTATOR_PLUGIN_STRATEGY.name,
  strategy.MUTATOR_PLUGIN_RADAMSA_STRATEGY.name,
]

def has_existing_mutator_strategy(fuzzing_strategies):
  return any(strategy in fuzzing_strategies for strategy in MUTATOR_STRATEGIES)

...

if (not has_existing_mutator_strategy(fuzzing_strategies) and 
    strategy_pool.do_strategy(..)):

...

File: src/python/bot/fuzzers/options.py (hunk 34 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/options.py]: Done

File: src/python/bot/fuzzers/libfuzzer.py (hunk 35 above)
Comment (mpherman2) [reply src/python/bot/fuzzers/libfuzzer.py]: Done
